File: mobile/lib/features/swims/domain/swim.dart
/// A practice swim assigned to this timer, as returned by `GET /my-swims`.
class Swim {
  const Swim({
    required this.id,
    required this.swimmerName,
    required this.venueName,
    required this.laneNo,
    required this.scheduledStart,
    required this.status,
    required this.lapTarget,
    required this.assignedTimerId,
  });

  final int id;
  final String swimmerName;
  final String venueName;
  final int laneNo;
  final DateTime scheduledStart;
  final String status;
  final int? lapTarget;
  final int? assignedTimerId;

  bool get isLive => status == 'live';
  bool get isScheduled => status == 'scheduled';

  factory Swim.fromJson(Map<String, dynamic> j) => Swim(
        id: j['id'] as int,
        swimmerName: (j['swimmer_name'] as String?) ?? 'Swimmer',
        venueName: (j['venue_name'] as String?) ?? '',
        laneNo: j['lane_no'] as int,
        scheduledStart: DateTime.parse(j['scheduled_start'] as String).toLocal(),
        status: j['status'] as String,
        lapTarget: j['lap_target'] as int?,
        assignedTimerId: j['assigned_timer_id'] as int?,
      );
}

File: backend/app/models/swim.py
"""Swim ORM model — one swimmer's practice swim: one swimmer, one lane, one timer, N laps."""
from __future__ import annotations

from datetime import datetime

from sqlalchemy import DateTime, ForeignKey, Integer, String
from sqlalchemy.orm import Mapped, mapped_column

from app.db.session import Base


class SwimRow(Base):
    __tablename__ = "swims"

    id: Mapped[int] = mapped_column(primary_key=True)
    venue_id: Mapped[int] = mapped_column(ForeignKey("venues.id"), nullable=False, index=True)
    swimmer_id: Mapped[int] = mapped_column(ForeignKey("swimmers.id"), nullable=False, index=True)
    lane_no: Mapped[int] = mapped_column(Integer, nullable=False)
    scheduled_start: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    lap_target: Mapped[int | None] = mapped_column(Integer, nullable=True)
    status: Mapped[str] = mapped_column(String(20), nullable=False, default="scheduled", index=True)
    closed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    closure_method: Mapped[str | None] = mapped_column(String(30), nullable=True)
